keps/sig-api-machinery/customresources-pruning-and-defaulting.md

```diff
@@ -0,0 +1,486 @@
+
+## Goals
+
+* Prune unknown fields from CustomResources silently. Unknown means not to be specified in the OpenAPI validation spec.
+* Allow to opt-out of pruning via the OpenAPI validation spec for a subtree of the JSON objects.
+* Apply OpenAPI validation schema defaults to CustomResources.
+* Have simple semantics for pruning and defaulting.
+
+## Non-Goals
+* Add a strict mode to the REST API which rejects objects with unknown fields.
+* Propose or decide anything about ratcheting. [We discuss the topic](#ratcheting-(slightly-related)) because it is related, but this needs a follow-up KEP.
+
+## Motivation
+### Pruning
+Pruning of a JSON value means to remove "unknown fields". A field (given as a JSON path) is unknown if it is not specified in the OpenAPI validation schema.
+
+A JSON path `<JSON path>.x` is specified in an OpenAPI validation schema if `"x" is a key in a corresponding `properties` field in the schema.
+
+#### Example 1
+
+>Assume the OpenAPI schema
+>
+>```json
+>{"properties": {"a": {}, "b": {"type": "string"}, "c": {"not": {}}}}
+>```
+>
+>Then a JSON object `{"a":1, "c": 3, "d": 4}` is pruned to `{"a":1, "c":3}`.
+>
+> Note that the pruned object does not validate.
+
+#### Example 2
+
+> Assume the OpenAPI schema
+>
+>```json
+>{"anyOf": [{"properties": {"a": {}}}, {"properties": {"b": {}}}]}
+>```
+>
+>Then a JSON object `{"a":1, "b": 2, "c": 3}` is pruned to `{"a":1, "b":2}` with the given semantics. Note that also `{"a":1}`, `{"b":2}` would be natural pruning results if we defined "specified" in a different way.
+
+#### Example 3
+
+> Assume the OpenAPI schema
+>
+>```json
+>{
+>   "properties": {
+>       "a": {}, 
+>       "b": {"type": "string", "properties": {"x": {}}}
+>   }, 
+>   "anyOf": [
+>	    { "properties": {"c":{}} }, 
+>       { "properties": {"d":{"not":{}}}}
+>   ]
+>}
+>```
+>
+> Then `a`, `b`, `b.x`, `c`, `d` are specified, `b.y` and `e` are not specified.
+
+**Question:** is it natural semantics to assume `d` to be specified and not to prune it? Note that there is no object with `d` that validates against `"d":{not:{}}`. But due to the `anyOf` there are objects which will validate against the complete schema.
+
+#### Pruning Options
+
+Motivated by the examples, we have different options to define the pruning semantics:
+1. Use the described semantics of `specified`.
+2. Only consider `properties` fields in the schema which actually successfully validate a given object (then `d` would be pruned in example 3).
+3. Only consider `properties` fields outside of `anyOf`, `allOf`, `oneOf`, `not` during pruning.
+4. Only consider `properties` fields outside of `anyOf`, `allOf`, `oneOf`, `not` during pruning, but enforce that every `properties` key inside `anyOf`, `allOf`, `oneOf`, `not` also appears outside all of those.
```

Comment:
**Antoine Pelisse:** Isn't that the same as the one before? Since if you drop the fields in oneOf, allOf, anyOf, they are always going to be invalid, so they have to also appear in `properties`.

**Stefan Schimanski:** We would reject the CRD if the outer properties are not a super-set of the ones under all/any/oneOf/not.

So option 4 protects at least the user from mistakes.

Note: all these are not about validating a CR or not, but how to specifiy what you don't want to prune via CRD OpenAPI validation schemata.